src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs

```diff
@@ -41117,6 +41117,156 @@ static void Main()
+        var verifier = CompileAndVerify(comp, expectedOutput: IncludeExpectedOutput("2"), verify: ExecutionConditionUtil.IsMonoOrCoreClr ? Verification.Passes : Verification.Skipped).VerifyDiagnostics();
+
+        verifier.VerifyIL("E.Method(C)",
+"""
+{
+  // Code size       46 (0x2e)
+  .maxstack  3
+  IL_0000:  newobj     "E.<>c__DisplayClass0_0..ctor()"
+  IL_0005:  dup
+  IL_0006:  ldarg.0
+  IL_0007:  stfld      "C E.<>c__DisplayClass0_0.<>4__this"
+  IL_000c:  dup
+  IL_000d:  ldftn      "void E.<>c__DisplayClass0_0.<Method>b__0()"
+  IL_0013:  newobj     "System.Action..ctor(object, nint)"
+  IL_0018:  callvirt   "void System.Action.Invoke()"
+  IL_001d:  ldftn      "void E.<>c__DisplayClass0_0.<Method>b__1()"
+  IL_0023:  newobj     "System.Action..ctor(object, nint)"
+  IL_0028:  callvirt   "void System.Action.Invoke()"
+  IL_002d:  ret
+}
+""");
+
+        comp = CreateCompilation(src1 + src2, targetFramework: TargetFramework.Net80, options: TestOptions.DebugExe);
+        CompileAndVerify(comp, expectedOutput: IncludeExpectedOutput("2"), verify: ExecutionConditionUtil.IsMonoOrCoreClr ? Verification.Passes : Verification.Skipped).VerifyDiagnostics();
+    }
+
+    [Fact]
+    public void InstanceMethod_Metadata_23_LambdaCapturing()
+    {
+        var src1 = """
+public implicit extension E for C
+{
+    public void Method()
+    {
+        var d1 = () => this.Increment();
+        d1();
+        var d2 = () => this.EIncrement();
+        d2();
+    }
+
+    public void EIncrement()
+    {
+        this.Increment();
+    }
+}
+
+public struct C
+{
+    public int F;
+
+    public void Increment()
+    {
+        F++;
+    }
+}
+""";
+        var comp = CreateCompilation(src1, targetFramework: TargetFramework.Net80, options: TestOptions.ReleaseDll);
+        comp.VerifyDiagnostics(
+            // (5,24): error CS1673: Anonymous methods, lambda expressions, query expressions, and local functions inside structs cannot access instance members of 'this'. Consider copying 'this' to a local variable outside the anonymous method, lambda expression, query expression, or local function and using the local instead.
+            //         var d1 = () => this.Increment();
+            Diagnostic(ErrorCode.ERR_ThisStructNotInAnonMeth, "this").WithLocation(5, 24),
+            // (7,24): error CS1673: Anonymous methods, lambda expressions, query expressions, and local functions inside structs cannot access instance members of 'this'. Consider copying 'this' to a local variable outside the anonymous method, lambda expression, query expression, or local function and using the local instead.
+            //         var d2 = () => this.EIncrement();
+            Diagnostic(ErrorCode.ERR_ThisStructNotInAnonMeth, "this").WithLocation(7, 24)
+            );
+    }
+
+    [Fact]
+    public void InstanceMethod_Metadata_24_LambdaCapturing()
+    {
+        var src1 = """
+public implicit extension E<T> for T where T : I1
```

Comment:
What happens when when this is `extension E<T> for T where T : class, I1` (i.e., we know `T` is a reference type)?

Reply:
I think in this case the code would compile successfully. The error is based on RefKind of `this` parameter and it will be None for an extended type known to be a reference type.